fill.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package weaver
+
+import (
+	"fmt"
+	"go/token"
+	"net"
+	"reflect"
+
+	"github.com/ServiceWeaver/weaver/internal/reflection"
+	"github.com/ServiceWeaver/weaver/internal/weaver"
+	"golang.org/x/exp/slog"
+)
+
+func init() {
+	// See internal/weaver/types.go.
+	weaver.SetLogger = setLogger
+	weaver.FillRefs = fillRefs
+	weaver.FillListeners = fillListeners
+}
+
+// See internal/weaver/types.go.
+func setLogger(v any, logger *slog.Logger) error {
+	x, ok := v.(interface{ setLogger(*slog.Logger) })
+	if !ok {
+		return fmt.Errorf("FillLogger: %T does not implement weaver.Implements", v)
+	}
+	x.setLogger(logger)
+	return nil
+}
+
+// See internal/weaver/types.go.
+func fillRefs(impl any, get func(reflect.Type) (any, error)) error {
+	p := reflect.ValueOf(impl)
+	if p.Kind() != reflect.Pointer {
+		return fmt.Errorf("FillRefs: %T not a pointer", impl)
+	}
+	s := p.Elem()
+	if s.Kind() != reflect.Struct {
+		return fmt.Errorf("FillRefs: %T not a struct pointer", impl)
+	}
+
+	for i, n := 0, s.NumField(); i < n; i++ {
+		f := s.Field(i)
+		if !f.Type().Implements(reflection.Type[interface{ isRef() }]()) {
+			continue
+		}
+
+		// Sanity check that field type structure matches weaver.Ref[T].
+		if f.Kind() != reflect.Struct {
+			panic(fmt.Errorf("FillRefs: weaver.Ref %v is not a struct", f))
+		}
+		if f.NumField() != 1 {
+			panic(fmt.Errorf("FillRefs: weaver.Ref %v does not have one field", f))
+		}
+		if f.Type().Field(0).Name != "value" {
+			panic(fmt.Errorf("FillRefs: weaver.Ref %v field 0 not named %q", f, "value"))
+		}
+
+		// Set the component.
+		valueField := f.Field(0)
+		component, err := get(valueField.Type())
+		if err != nil {
+			return fmt.Errorf("FillRefs: setting field %v.%s: %w", s.Type(), s.Type().Field(i).Name, err)
+		}
+		setPossiblyUnexported(valueField, reflect.ValueOf(component))
+	}
+	return nil
+}
+
+// See internal/weaver/types.go.
+func fillListeners(impl any, get func(name string) (net.Listener, string, error)) error {
+	p := reflect.ValueOf(impl)
+	if p.Kind() != reflect.Pointer {
+		return fmt.Errorf("FillListeners: %T not a pointer", impl)
+	}
+	s := p.Elem()
+	if s.Kind() != reflect.Struct {
+		return fmt.Errorf("FillListeners: %T not a struct pointer", impl)
+	}
+
+	for i, n := 0, s.NumField(); i < n; i++ {
+		f := s.Field(i)
+		t := s.Type().Field(i)
+		if f.Type() != reflection.Type[Listener]() {
+			continue
+		}
+
+		// The listener's name is the field name, unless a tag is present.
+		name := t.Name
+		if tag := t.Tag.Get("weaver"); tag != "" {
+			if !token.IsIdentifier(tag) {
```

Comment:
Is token too big of a dependency to pull in just to be able to use IsIdentifier? Do we have an identifier checker somewhere else we can call instead?

You are just moving code for now, so no reason to change things in this PR, but I spotted the dep now and wanted to note the issue.

Reply:
Good point. Let me replace it in another PR. The implementation is dirt simple, so I'll just copy it over: https://cs.opensource.google/go/go/+/refs/tags/go1.20.6:src/go/token/token.go;l=331-341